### PR TITLE
[WIP] Backport to ARMv7-A

### PR DESCRIPTION
**Pull request by easyaspi314**
With some small polyfills for the new Q-form instructions, we can easily support ARMv7-A. 

Makefile needs to be updated, and I am not sure if we should change the paths/namespaces.

Tests seem to be passing on a Snapdragon 835 in 32-bit mode.

### CODE REVIEW

**Comment by lemire** [comment]
Wow. This is surprisingly thin as far as implementations are concerned. :-)

**Comment by lemire** [comment]
Your fork is not recognized by GitHub as a fork, so it appears that I cannot issue a PR to it from a branch in this repo (lemire/simdjson).

Can you consider merging the following branch...

https://github.com/lemire/simdjson/tree/armv7-ci-testings

The idea is as follows... We use the 32-bit drone.io instances (which are armv8) where we compile the binary to armv7. This should allow us to test armv7 since the 32-bit armv8 is effectively able to run armv7, I believe.

Short of that, we might be able to test with 32-bit armv8.


In any case, we can't possible merge your PR if we don't have some sensible way to test it automatically. I think that it is our best bet right now.

**Comment by lemire** [comment]
This interacts somewhat with @jkeiser's fallback architecture.

**Comment by jkeiser** [comment]
@lemire what do we want to do with this? I'm not sure how to test it, but it'd be nice to add the support if it really works :)

**Comment by lemire** [comment]
>  what do we want to do with this? I'm not sure how to test it, but it'd be nice to add the support if it really works :)

The first step would be figuring out whether it has performance benefits over the fallback implementation. That requires access to a good 32-bit ARM system.

Anyone wants to run benchmarks over an old Raspberry Pi?

**Comment by jkeiser** [comment]
I'll dust mine off and see what version it is.

**Comment by jkeiser** [comment]
We should put this to rest. Either:
* Close it and create an issue so it doesn"t get forgotten
* Merge it or something like it, and mark arm v7 support experimental / untested.

I'm leaning towards the first since we don't have evidence this has been tested and also don't know how it performs (even though it might be awesome!).

**Comment by lemire** [comment]
We can close it.

**Comment by jkeiser** [comment]
Closing after filing #1508, but please don't hesitate to reopen if you've got the time! One possible way forward, mentioned in #1508, is to add a #define SIMDJSON_UNSUPPORTED_ARCHITECTURES so people could bring in ARMv7 even though it's not in CI. At least we'd have it in the codebase then!

**Comment by jkeiser** [comment]
Heck, even turning SIMDJSON_IMPLEMENTATION_ARMV7 off by default (unless the user explicitly enables it) would be enough, I think.

File: include/simdjson/portability.h [hunk 1]
```diff
@@ -4,7 +4,7 @@
 #if defined(__x86_64__) || defined(_M_AMD64)
 #define IS_X86_64 1
 #endif
-#if defined(__aarch64__) || defined(_M_ARM64)
+#if defined(__aarch64__) || defined(_M_ARM64) || defined(__ARM_NEON__) || defined(__ARM_NEON)
 #define IS_ARM64 1
```
Comment (pps83) [inline include/simdjson/portability.h:8]: this should be reworked, it's confusing to say IS_ARM64 when building arm32. Should be renamed to IS_ARM and those arm64 polyfills should be defined for regular arm.

Comment (lemire) [reply include/simdjson/portability.h:8]: This PR was dropped. It was a proof of concept and should be reworked.

**Pull request closed without merging**